### DISCUSSION TITLE
Using/type setting agent/instance attributes for arithmetics in wealth model

### DISCUSSION THREAD

**Discussion by yannickoswald** [JoelForamitti/agentpy#39] [Q&A] [answered]
Hi all, Hi Joel, 

well done. Super work and documentation. Super efficient code and everything.  I am currently trying to teach myself OOP and in parallel agentpy. 

First newbie problem I seem to not overcome is with the def setup() function. I wonder why not use the __init__ method? Like Mesa does , is there a specific reason? 

My specific issue I would like to change the rules of the wealth transfer model from

         def wealth_transfer(self):
        
                 if self.wealth > 0:
        
                    partner = self.model.agents.random()
                    partner.wealth += 1 
                    self.wealth -= 1 * partner
    
to something like this (just for learning purposes)

          def wealth_transfer(self):

                 if self.wealth > 0:
        
                    partner = self.model.agents.random()
                    wealth_to_divide  = partner.wealth + self.wealth
                    rule_to_divide = self.wealth/(partner.wealth + self.wealth)
                    partner.wealth +=  wealth_to_divide * (1- rule_to_divide)
                    self.wealth +=wealth_to_divide *rule_to_divide
       `

It gives me the error 

TypeError: unsupported operand type(s) for *: 'int' and 'AgentIter'

It does not seem to allow the arithmetic useing the instance attr. `partner.wealth`

I tried to somehow typeset the attribute but did not succeed so far.

However in Mesa with the __init__ method it works without issues. I just wonder what's the difference?

Thanks a lot!

Yannick

**Comment by jofmi** [comment, marked as answer]
Hi Yannik,

glad you like the package!

regarding `setup()`, you can also overwrite `__init__()` instead. If you look at the source, the agent's `__init__()` looks just like this:

```
def __init__(self, model, *args, **kwargs):
    super().__init__(model)
    self.setup(*args, **kwargs)
```

regarding your problem `self.model.agents.random()` returns an iterator of agents (even if in your case it is only one agent). the iterator can support some operations but not all. an easy way to unpack the agent from the iterator is this (notice the comma after partner):

```
partner, = self.model.agents.random()
```

if you want to be more explicit, you can also use:

```
for partner in self.model.agents.random():
   ...
```

let me know if that solves the problem. I should add that info to the documentation, then

Best,
Joël

**Reply by yannickoswald** [reply]
Great thanks a lot Joël for the swift reply. That works! :). And very nice the code remains so concise
